ray-curator/ray_curator/stages/deduplication/semantic/kmeans.py

```diff
@@ -0,0 +1,371 @@
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Literal
+
+import cupy as cp
+import numpy as np
+
+from ray_curator.backends.base import WorkerMetadata
+from ray_curator.stages.base import CompositeStage, ProcessingStage
+from ray_curator.stages.deduplication.io_utils import DeduplicationIO
+from ray_curator.stages.file_partitioning import FilePartitioningStage
+from ray_curator.stages.resources import Resources
+from ray_curator.stages.text.embedders.utils import create_list_series_from_1d_or_2d_ar
+from ray_curator.tasks import FileGroupTask, _EmptyTask
+
+from .utils import break_parquet_partition_into_groups, get_array_from_df
+
+if TYPE_CHECKING:
+    import cudf
+
+import time
+
+import torch
+from loguru import logger
+
+# Column names
+L2_DIST_TO_CENT_COL = "l2_dist_to_cent"
+COSINE_DIST_TO_CENT_COL = "cosine_dist_to_cent"
+
+FILETYPE_TO_EXTENSIONS = {
+    "parquet": [".parquet"],
+    "jsonl": [".jsonl", ".json"],
+}
+# Create a flat extension -> filetype mapping
+EXTENSION_TO_FILETYPE = {ext: filetype for filetype, exts in FILETYPE_TO_EXTENSIONS.items() for ext in exts}
```

Comment:
Create a util for this and put it in file utils please , we will all use from this and ensure consitent matching

Reply:
This is where KMeans (and maybe dedup specific) so don't want to move it up, because everywhere else we use FilePartitioningStage which accepts `file_extensions` while KMeans accepts `filetype` (since the I/O is different).